### PR TITLE
Added support for the `kotlin.Nothing` class as built-in

### LINKED ISSUE #614: Serializer for Nothing not found
**Describe the bug**
Using `Nothing` in a generic type argument fails with an internal error `Serializer for element of type Nothing has not been found`.

**To Reproduce**
This reproduces the problem when compiling:
```kotlin
import kotlinx.serialization.Serializable

@Serializable
sealed class SerializableEither<out L, out R> {

    @Serializable
    data class Left<L>(val value: L) : SerializableEither<L, Nothing>()

    @Serializable
    data class Right<R>(val value: R) : SerializableEither<Nothing, R>()
}
```

Full compiler error: https://pastebin.com/FP6zGrDf

**Expected behavior**
`Nothing` should be serializable.

**Environment**
 - Kotlin version: 1.3.60
 - Library version: 0.14.0
 - Kotlin platforms: JVM
 - Gradle version: 5.6.4
 - IDE version (if bug is related to the IDE): IntelliJ IDEA 2019.2.4 (Ultimate Edition)
- Runtime version: 11.0.4+10-b304.77 amd64
- Linux 5.0.0-36-generic

### LINKED ISSUE #932: Serializer for Nothing
A `KSerializer<Nothing?>` should serve the same purposes that `Nothing` in type system: indicate that a particular item should never be serialized. There are several use-cases for it, mainly connected with generics:
https://github.com/Kotlin/kotlinx.serialization/issues/607
https://github.com/Kotlin/kotlinx.serialization/issues/393
https://github.com/Kotlin/kotlinx.serialization/issues/614

**Describe the solution you'd like**
A `NothingSerializer` that can be used in `@Serializable(NothingSerializer)` on types. Or `NoSerializer` serializer. Or `@NoSerializer` annotation as a shorthand.

### PR DESCRIPTION
**Pull request by shanshin**
Resolves #614
Resolves #932

### CODE REVIEW

**Review by sandwwraith — CHANGES_REQUESTED** [review]

File: docs/builtin-classes.md [hunk 1]
```diff
@@ -380,6 +381,33 @@ which is explained in the [Polymorphism. Objects](polymorphism.md#objects) secti
 
 > Serialization of objects is format specific. Other formats may represent objects differently, 
 > e.g. using their fully-qualified names.
+
+## Nothing
+
+By default, [Nothing] is a serializable class. However, since there are no instances of this class, it is impossible to encode or decode its values - any attempt will cause an exception.
+
+This serializer is used when syntactically is needed some type, but it is not actually used in serialization. For example, when using parameterized classes:
```
[the remark this answers is not in the document]
Comment (shanshin) [reply docs/builtin-classes.md]: Thanks!

**Review by sandwwraith — APPROVED** [review]

**Comment by qwwdfsad** [comment]
Waiting for 1.8.0

**Comment by nicolasfara** [comment]
@qwwdfsad Kotlin 1.8.0 is out. Any news about the merge of this PR?

**Comment by sandwwraith** [comment]
@shanshin #2146 is merged, you can rebase this PR

File: core/commonMain/src/kotlinx/serialization/internal/BuiltInSerializers.kt [hunk 2]
```diff
@@ -23,3 +24,16 @@ internal object DurationSerializer : KSerializer<Duration> {
         return Duration.parseIsoString(decoder.decodeString())
     }
 }
+
+@PublishedApi
+internal object NothingSerializer : KSerializer<Nothing> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("kotlin.Nothing", PrimitiveKind.INT)
```
Comment (sandwwraith) [inline core/commonMain/src/kotlinx/serialization/internal/BuiltInSerializers.kt:30]: Have we discussed this design? I'm not sure reusing PrimitiveKind.INT is good idea

File: formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt [hunk 3]
```diff
@@ -279,6 +279,13 @@ class SerializersLookupTest : JsonTestBase() {
         }
     }
 
+// TODO uncomment when Kotlin 1.7.20 is released
```
Comment (sandwwraith) [inline formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt]: Should this be uncommented?

Comment (shanshin) [reply formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt]: Not actual code

File: core/commonMain/src/kotlinx/serialization/internal/BuiltInSerializers.kt (hunk 2 above)
Comment (shanshin) [reply core/commonMain/src/kotlinx/serialization/internal/BuiltInSerializers.kt:30]: New primitive kind?

File: formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt (hunk 3 above)
Comment (shanshin) [reply formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt]: This change wasn't include in the merged commit.

File: core/commonMain/src/kotlinx/serialization/internal/BuiltInSerializers.kt (hunk 2 above)
Comment (shanshin) [reply core/commonMain/src/kotlinx/serialization/internal/BuiltInSerializers.kt:30]: #2150

File: formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt (hunk 3 above)
Comment (shanshin) [reply formats/json-tests/commonTest/src/kotlinx/serialization/SerializersLookupTest.kt]: https://github.com/Kotlin/kotlinx.serialization/pull/1991/commits/a5d92ec747290199d7828883bd71b6b4d107e6ab

**Pull request merged**